File: 22.02/1654_랜선자르기.py
import sys
input = sys.stdin.readline

N, K = map(int, input().split())
lst = []
for i in range(N):
    lst.append(int(input()))
E = max(lst)

start = 1
end = E
max_v = 0
while start <= end:
    middle = (start + end) // 2

    tmp = 0
    for i in lst:
        tmp += (i // middle)
    if tmp >= K and middle > max_v:
        max_v = middle

    if tmp < K:
        end = middle - 1

    elif tmp >= K:
        start = middle + 1

print(max_v)

File: 24.03/1654_랜선자르기.js
const input = require("fs")
  .readFileSync("./input.txt") // .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [K, N] = input[0].split(" ").map(Number);
const lans = input.splice(1).map(Number);

let start = 1;
let end = Math.max(...lans);
let answer = 0;

while (start <= end) {
  let middle = parseInt((start + end) / 2);
  let tmp = 0;
  for (let i of lans) {
    tmp += parseInt(i / middle);
  }

  if (tmp >= N && middle > answer) answer = middle;

  if (tmp < N) end = middle - 1;
  else if (tmp >= N) start = middle + 1;
}

console.log(answer);
process.exit();